src/runtime/opencl/memory_pool.cc

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file memory_pool.h
+ * \brief OpenCL Workspace pool utility.
+ */
+#include "memory_pool.h"
+
+#include <memory>
+
+#include "opencl_common.h"
+
+#define ALIGN_UP(num, align) (((num) + ((align)-1)) & ~((align)-1))
+
+namespace tvm {
+namespace runtime {
+namespace cl {
+
+void* Pool::Alloc(Device dev, DeviceAPI* device, size_t nbytes) {
+  Entry e;
+  DLDataType type;
+  type.code = kDLUInt;
+  type.bits = 8;
+  type.lanes = 1;
+  if (free_list_.size() == 2) {
+    e = free_list_.back();
+    free_list_.pop_back();
+    if (e.size < nbytes) {
+      // resize the page
+      OpenCLWorkspace::Global()->FreeCLBuffer(dev, e.data);
+      e.data = OpenCLWorkspace::Global()->AllocCLBuffer(dev, nbytes, kTempAllocaAlignment, type);
+      e.size = nbytes;
```

Comment:
Reading the set of changes, I feel we are pushing the memory management logic a bit too into the DeviceAPI and NDArray. I wonder if in this case what we want is an alternative memory allocator.